core/src/impl/Kokkos_Profiling.hpp

```diff
@@ -135,6 +137,69 @@ Kokkos_Profiling_SpaceHandle make_space_handle(const char* space_name);
 
 namespace Experimental {
 
+namespace Impl {
+struct DirectFenceIDHandle {
+  uint32_t value;
+};
+//
+template <typename Space>
+uint32_t idForInstance(const uintptr_t instance) {
+  /** Needed to be a ptr due to initialization order problems*/
+  using map_type = std::map<uintptr_t, uint32_t>;
+
+  static std::shared_ptr<map_type> map;
+  if (map == nullptr) {
+    map = std::make_shared<map_type>();
+  }
+
+  static uint32_t value;
+  constexpr const uint32_t offset =
+      Kokkos::Tools::Experimental::NumReservedDeviceIDs;
+
+  auto find = map->find(instance);
+  if (find == map->end()) {
+    auto ret         = offset + value++;
```

Comment:
In time, probably? But as of now I don't think we're requiring thread-safety.

Reply:
actually this needs to be thread safe right now since you call this without a tool being loaded, and stuff like the CUDA backend is used in a threaded mode already.